File: backend/app/science/export.py
"""Generic ICRS RA/DEC CSV export for enabled proposed tile centers."""

from __future__ import annotations

import csv
import io

from app.models import ExportRequest, TileSource
from app.profiles import load_profile
from app.science.coordinates import format_dec_degrees, format_ra_degrees


def build_export_csv(request: ExportRequest) -> str:
    """Serialize active proposed ICRS centers to interoperable CSV.

    Parameters
    ----------
    request : ExportRequest
        Proposed RA/DEC centers in degrees, a profile-approved descriptive
        epoch label, and decimal-degree or sexagesimal representation.

    Returns
    -------
    str
        UTF-8-ready CSV with ``RA,DEC,EPOCH``. Decimal coordinates use eight
        fractional degree digits; sexagesimal RA uses hours and DEC degrees.

    Raises
    ------
    ValueError
        If there is no active proposed tile, a non-proposal is supplied, or
        the epoch is not permitted by the active profile.

    Notes
    -----
    The EPOCH value is catalogue metadata. This function does not precess
    coordinates, change frame/equinox, or apply proper motion.
    """
    profile = load_profile(request.profile_id)
    epoch = request.epoch if request.epoch is not None else profile.export_epoch_default
    if epoch not in profile.export_epoch_options:
        raise ValueError(f"Epoch {epoch!r} is not allowed by profile {profile.id}")
    if any(tile.source != TileSource.PROPOSED for tile in request.proposed_tiles):
        raise ValueError("Generic export accepts only proposed tile centers")
    active = [tile for tile in request.proposed_tiles if tile.enabled]
    if not active:
        raise ValueError("Enable at least one proposed tile before downloading")

    output = io.StringIO(newline="")
    writer = csv.DictWriter(output, fieldnames=("RA", "DEC", "EPOCH"), lineterminator="\r\n")
    writer.writeheader()
    for tile in active:
        if request.coordinate_format == "decimal":
            ra, dec = f"{tile.ra_deg:.8f}", f"{tile.dec_deg:.8f}"
        else:
            ra = format_ra_degrees(tile.ra_deg, precision=3)
            dec = format_dec_degrees(tile.dec_deg, precision=3)
        writer.writerow({"RA": ra, "DEC": dec, "EPOCH": epoch})
    return output.getvalue()

File: frontend/src/science/export.ts
import type { CoordinateFormat, TileRecord, TilingProfile } from "../types";
import { formatDecDegrees, formatRaDegrees } from "./coordinates";
import { resolveProfile } from "../profiles";

/** Serialize active proposed ICRS centers to the Python-compatible generic CSV.
 * @param proposedTiles - Up to 500 proposals with decimal-degree ICRS centers.
 * @param profileId - Installed profile ID or custom.
 * @param epoch - Descriptive catalogue epoch label; no precession is applied.
 * @param coordinateFormat - Decimal degrees or sexagesimal hour-angle RA and degree DEC.
 * @param inlineProfile - Session-only custom geometry when profileId is custom.
 * @returns UTF-8-ready CSV text with CRLF and RA,DEC,EPOCH fields.
 * @throws On invalid tiles, format, epoch, profile, or no enabled proposal.
 */
export function buildExportCsv(proposedTiles: TileRecord[], profileId = "splus-t80-south", epoch?: string, coordinateFormat: CoordinateFormat = "decimal", inlineProfile?: TilingProfile): string {
  if (proposedTiles.length > 500) throw new Error("Too many proposed tiles");
  const profile = resolveProfile(profileId, inlineProfile);
  const selectedEpoch = epoch ?? profile.export_epoch_default;
  if (!profile.export_epoch_options.includes(selectedEpoch)) throw new Error(`Epoch '${selectedEpoch}' is not allowed by profile ${profile.id}`);
  if (proposedTiles.some((tile) => tile.source !== "proposed")) throw new Error("Generic export accepts only proposed tile centers");
  if (!["decimal", "sexagesimal"].includes(coordinateFormat)) throw new Error("Invalid coordinate format");
  const active = proposedTiles.filter((tile) => tile.enabled !== false);
  if (!active.length) throw new Error("Enable at least one proposed tile before downloading");
  const rows = active.map((tile) => {
    if (!Number.isFinite(tile.ra_deg) || tile.ra_deg < 0 || tile.ra_deg >= 360 || !Number.isFinite(tile.dec_deg) || Math.abs(tile.dec_deg) > 90 || !tile.generation_method) throw new Error(`Invalid proposed tile ${tile.id}`);
    const ra = coordinateFormat === "decimal" ? tile.ra_deg.toFixed(8) : formatRaDegrees(tile.ra_deg, 3);
    const dec = coordinateFormat === "decimal" ? tile.dec_deg.toFixed(8) : formatDecDegrees(tile.dec_deg, 3);
    return `${ra},${dec},${selectedEpoch}`;
  });
  return ["RA,DEC,EPOCH", ...rows, ""].join("\r\n");
}
